Add blogpost for new Prometheus connector

nicusX opened the pull request:
Add new blogpost introducing Prometheus connector.


Blog preview (updated): 
[prometheus-blog-preview.pdf](https://github.com/user-attachments/files/18024453/prometheus-blog-preview.pdf)

hlteoh37 commented on docs/content/posts/2024-11-26-introducing-new-prometheus-connector.md:
```diff
@@ -0,0 +1,201 @@
+---
+
+
+We are excited to announce a new sink connector that enables writing data to Prometheus ([FLIP-312](https://cwiki.apache.org/confluence/display/FLINK/FLIP-312:+Prometheus+Sink+Connector)). This articles introduces the main features of the connector, the reasoning behind design decisions.
```
```suggestion
We are excited to announce a new sink connector that enables writing data to Prometheus ([FLIP-312](https://cwiki.apache.org/confluence/display/FLINK/FLIP-312:+Prometheus+Sink+Connector)). This articles introduces the main features of the connector, and the reasoning behind design decisions.
```

hlteoh37 commented on docs/content/posts/2024-11-26-introducing-new-prometheus-connector.md:
```diff
@@ -0,0 +1,201 @@
+
+Prometheus is an efficient time-series database optimized for building real-time dashboards and alerts, typically in combination with Grafana or other visualization tools.
+
+Prometheus is commonly used to monitor compute resources, IT infrastructure, Kubernetes clusters, applications, and cloud resources. It can also be used to observe your Flink cluster and Flink jobs. Flink already has [Metric Reporters](https://nightlies.apache.org/flink/flink-docs-master/docs/deployment/metric_reporters/) to export metrics to Prometheus. 
```
```suggestion
Prometheus is commonly used to monitor compute resources, IT infrastructure, Kubernetes clusters, applications, and cloud resources. It can also be used to observe your Flink cluster and Flink jobs. Flink has existing [Metric Reporters](https://nightlies.apache.org/flink/flink-docs-master/docs/deployment/metric_reporters/) to export metrics to Prometheus for this purpose. 
```

nicusX replied:
Not sure the new sentence works. I am changing to:
> Prometheus is commonly used to monitor compute resources, IT infrastructure, Kubernetes clusters, applications, and cloud resources. It can also be used to observe your Flink cluster and Flink jobs. Flink **existing** [Metric Reporters](https://nightlies.apache.org/flink/flink-docs-master/docs/deployment/metric_reporters/) **has this purpose**.

hlteoh37 commented on docs/content/posts/2024-11-26-introducing-new-prometheus-connector.md:
```diff
@@ -0,0 +1,201 @@
+
+So, why do we need a connector?
+
+Prometheus can serve as a general-purpose observability time-series database, beyond traditional infrastructure monitoring. For example, it can be used to monitor IoT devices, sensors, connected cars, media streaming devices, and any resource that streams events or measurements continuously.
+
+Observability data from these use cases differs from metrics generated by compute resources. Events are pushed by devices instead of being scraped, resulting in irregular frequency. Devices may be connected via mobile networks or even Bluetooth, causing events from each device to follow different paths and arrive at different times. The frequency and cardinality of events emitted by these devices ...
```
Wonder if we could/should rephrase to make the benefits clearer. e.g.


* **Events re-ordering**: Devices may be connected via mobile networks or even Bluetooth, causing events from each device to follow different paths and arrive at different times.
* **Reduce cardinality**: The frequency and cardinality of events emitted by these devices can be very high, making it challenging to derive insights directly.
* **Event Enrichment**: Events often lack contextual information and require enrichment to add additional dimensions before being sent to a time-series database.

nicusX replied:
This is actually mixing a challenges and the solutions. I will rephrase into bulletpoints, listing the challenges, and what you need to do to overcome them

hlteoh37 commented on docs/content/posts/2024-11-26-introducing-new-prometheus-connector.md:
```diff
@@ -0,0 +1,201 @@
+
+Observability data from these use cases differs from metrics generated by compute resources. Events are pushed by devices instead of being scraped, resulting in irregular frequency. Devices may be connected via mobile networks or even Bluetooth, causing events from each device to follow different paths and arrive at different times. The frequency and cardinality of events emitted by these devices ...
+
+### Flink as observability events pre-processor
+
+You can address the challenges above by pre-processing raw events with Flink. Aggregation over short time windows can reduce frequency and cardinality, filtering can remove noise, and enrichment with reference data can add additional context. These are all tasks that Flink can perform efficiently, and at scale.
+
+What has been missing until now is an easy, reliable way to write from Flink to Prometheus at scale.
+
```
This might be better merged into above section!

nicusX replied:
Got your point. I reorganised the paragraph

hlteoh37 commented on docs/content/posts/2024-11-26-introducing-new-prometheus-connector.md:
```diff
@@ -0,0 +1,201 @@
+
+### Non-trivial implementing an efficient Remote-Write client
+
+You could implement a sink from scratch or use AsyncIO to call the Prometheus Remote-Write HTTP endpoint, but you would need to manage all aspects yourself. Prometheus Remote-Write has no high-level client, so you would need to build on top of a low-level HTTP client. Additionally, Remote-Write can be inefficient unless writes are batched and parallelized. Error handling can be complex, and speci ...
```
Maybe we can add to the bold/bulleted list!

e.g.

* **Non-trivial implementation of efficient Remote-Write Client**. <Details>

nicusX replied:
Turning into bullet points

hlteoh37 reviewed (changes requested):
Added some comments!

nicusX commented:
@hlteoh37 addressed the comments. Also, updated the preview

dannycranmer commented on docs/content/posts/2024-11-26-introducing-new-prometheus-connector.md:
```diff
@@ -0,0 +1,202 @@
+
+Prometheus is an efficient time-series database optimized for building real-time dashboards and alerts, typically in combination with Grafana or other visualization tools.
+
+Prometheus is commonly used to monitor compute resources, IT infrastructure, Kubernetes clusters, applications, and cloud resources. It can also be used to observe your Flink cluster and Flink jobs. Flink existing [Metric Reporters](https://nightlies.apache.org/flink/flink-docs-master/docs/deployment/metric_reporters/) has this purpose. 
```
> Flink existing [Metric Reporters](https://nightlies.apache.org/flink/flink-docs master/docs/deployment/metric_reporters/) has this purpose. 

This sentence feels orphaned and does not read right, not sure how it connects to the previous. I think I know what you mean, but can we improve?

dannycranmer commented on docs/content/posts/2024-11-26-introducing-new-prometheus-connector.md:
```diff
@@ -0,0 +1,202 @@
+* Remote-Write can be inefficient unless write requests are batched and parallelized.
+* Error handling can be complex, and specifications demand strict behaviors (see [Strict Specifications, Lenient Implementations](#strict-specifications-lenient-implementations)).
+
+The new Prometheus connector manages all of this for you.
```
I think in this section we should answer why/when we should use a connector rather than registering a metric in Flink.

nicusX replied:
Good point. I will try to clarify, hoping not to be too verbose.
But the bottomline is that this is **data** from the point of view of the application

dannycranmer commented on docs/content/posts/2024-11-26-introducing-new-prometheus-connector.md:
```diff
@@ -0,0 +1,202 @@
+* DataStream API, Java Sink, based on AsyncSinkBase.
+* Configurable write batching.
+* Order is retained on retries.
+* At-most-once guarantees (we’ll explain the reason for this later).
```
> At-most-once guarantees

Is this a feature or a limitation :o

nicusX replied:
Prometheus is at-most-once by design: fast ingestion and data freshness over consistency. If you try to make a Prometheus sink at-least-once, Prometheus will spit in your face :D 
If you make the connector fail when Prometheus rejects your data, you are doomed into an endless loop of restart-from-checkpoint

dannycranmer replied:
> Prometheus will spit in your face :D 

:D

dannycranmer commented on docs/content/posts/2024-11-26-introducing-new-prometheus-connector.md:
```diff
@@ -0,0 +1,202 @@
+* At-most-once guarantees (we’ll explain the reason for this later).
+* Supports parallelism greater than 1.
+* Configurable retry strategy for retryable server responses (e.g., throttling) and configurable behavior when the retry limit is reached.
+* Configurable behavior on retryable errors (e.g., throttling from the Prometheus backend).
```
Is this the same as previous line?

nicusX replied:
Yes! removed

dannycranmer commented on docs/content/posts/2024-11-26-introducing-new-prometheus-connector.md:
```diff
@@ -0,0 +1,202 @@
+
+Flink is designed with a consistency-first approach. By default, any unexpected error causes the streaming job to fail and restart from the last checkpoint.
+
+In contrast, Prometheus is designed with an availability-first approach, prioritizing fast ingestion over strict consistency. When a write request contains malformed entries, the entire request must be discarded and not retried. Additionally, samples belonging to the same time series (with the same dimensions or labels) must be written in strict timestamp order.
```
> the entire request must be discarded and not retried

Why can't we retry? I see the docs say we cannot retry 400 (beside 429), but what happens if you do?

nicusX replied:
Prometheus keeps rejecting. 
Because 400 means "malformed data" (for any reasons)

nicusX replied:
But good point. Adding a clarification about this

dannycranmer commented on docs/content/posts/2024-11-26-introducing-new-prometheus-connector.md:
```diff
@@ -0,0 +1,202 @@
+
+In contrast, Prometheus is designed with an availability-first approach, prioritizing fast ingestion over strict consistency. When a write request contains malformed entries, the entire request must be discarded and not retried. Additionally, samples belonging to the same time series (with the same dimensions or labels) must be written in strict timestamp order.
+
+You may have already spotted the issue: any malformed or out-of-order samples can act as a “poison pill” unless you drop the offending request and proceed.
```
How does this differ from a malformed write to any other sink? Is it because AsyncSink does not guarantee ordering?

nicusX replied:
Prometheus treats malformed and out-of-order the same way: 400 (rejected).
This sink has fixed in-flight request = 1 to prevent causing out-of-order while retrying

dannycranmer commented on docs/content/posts/2024-11-26-introducing-new-prometheus-connector.md:
```diff
@@ -0,0 +1,202 @@
+
+### Error-handling and guarantees
+
+As we have seen, the Remote-Write specifications require that certain rejected writes—specifically those related to malformed and out-of-order samples (typically 400 responses)—must not be retried. The connector has no choice but to discard the entire offending request and continue with the next record. The specifications do not define a standard response format, and there is no reliable way to a ...
```
Would the connector send out of order samples? Or is that determined by the user records?

nicusX replied:
The connector does not generate out-of-orderness itself.
For parallelism, we provide a KeySelector to partition the stream of input objects so that the same time-series are written by the same subtask. 
Upstream of that, it's responsibility of the user to correctly partition data to prevent accidental out-of-orderness. We made it very clear in the docs as well

nicusX replied:
Let me clarify this

dannycranmer commented on docs/content/posts/2024-11-26-introducing-new-prometheus-connector.md:
```diff
@@ -0,0 +1,202 @@
+
+For this reason, this is an *at-most-once* connector.
+
+If data loss is unavoidable, the best the connector can do is to make data loss very evident to the user. Each time a request is dropped, a `WARN` log is emitted, containing the endpoint’s response (which provides an indication of the issue) and incrementing counters for the number of dropped samples and requests.
```
What happens when we restart from a checkpoint? Would we discard a bunch of writes until we catch up?

nicusX replied:
Exactly that. Good point. Let me make this explicit

dannycranmer commented on docs/content/posts/2024-11-26-introducing-new-prometheus-connector.md:
```diff
@@ -0,0 +1,202 @@
+
+There are two ways to scale write throughput to Prometheus: batching multiple samples into a single write request and parallelizing the writes.
+
+Parallelizing poses challenges due to the ordering requirements imposed by Prometheus. You must ensure that all samples belonging to the same time series (defined as records with an identical set of labels) are written by the same thread.
```
Do we use Async Sink with 1 max inflight request?

nicusX replied:
Exactly. Let me add the clarification

dannycranmer commented on docs/content/posts/2024-11-26-introducing-new-prometheus-connector.md:
```diff
@@ -0,0 +1,202 @@
+
+To facilitate testing we created a data generator, a separate Flink application, capable of generating semi-random data and of introducing specific errors, like out-of-order samples. The generator can also produce a high volume of data for load and stress testing.  One aspect of the test harness requiring special attention was not loosing ordering. For example, we used Kafka between the generator ...
+
+Finally, the connector was also stress-tested, writing up to 1,000,000 samples per second with 1,000,000 cardinality (distinct time-series). 1 million is where we stopped testing, not a limit of the connector itself.
```
What parallelism was this at?

nicusX replied:
Max 32. Let me add this

nicusX commented:
@hlteoh37 @dannycranmer I addressed all your comments

nicusX commented:
@hlteoh37 @dannycranmer I addressed all your comments.

dannycranmer reviewed (approved):
LGTM, thanks @nicusX

Commit:
Add blogpost for new Prometheus connector
Addressed comments by @dannycranmer
Addressed comments by @hlteoh37
Update publication date to 2024-12-05T18:00:00.000Z

hlteoh37 reviewed (approved):
LGTM!
